### DISCUSSION TITLE
Monitoring Listener Performance with multiple MeterRegistry beans

### DISCUSSION THREAD

**Discussion by alexanderad** [spring-projects/spring-kafka#2105] [Q&A] [answered]
Hi!

Recently I've been looking for a way to monitor listener specific metrics and found a corresponding section in the docs ([Monitoring Listener Performance](https://docs.spring.io/spring-kafka/reference/html/#monitoring-listener-performance)) saying
> ... the listener container will automatically create and update Micrometer `Timer`s for the listener, **if Micrometer is detected** on the class path, **and a single MeterRegistry is present** in the application context ...

This way of collecting listener metrics is particulary useful given that it is able to account for multiple types of listeners (e.g. declared with `@KafkaListener`, or `@KafkaHandler`, or being batch listeners, or hidden behind custom factories for consumers, etc); as well as aware of concurrent listeners, and templates.

The corresponding implementation to be found in: https://github.com/spring-projects/spring-kafka/blob/7c2178ac4364989b84b9e644092d7dfb60882887/spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/MicrometerHolder.java#L61-L79

The problem I faced is that in case when multiple registries are auto-configured for the application context, e.g. Prometheus + Graphite timers for metrics do not register themselves.This applies to any combination of monitoring systems providing more that one `MeterRegistry.class` bean visible via `context.getBeansOfType(MeterRegistry.class, ...)`. `KafkaMessageListenerContainer` silently rejects to collect these metrics, and it seems there is no approachable way to configure alternative behavior, e.g.:
* prefer a specific registry (say, provided via `containerProperties`)
* combine discovered registries into [`CompositeMeterRegistry`](https://www.javadoc.io/doc/io.micrometer/micrometer-core/1.3.1/io/micrometer/core/instrument/composite/CompositeMeterRegistry.html)

I'm looking for an advice here. Is there anything obvious I'm missing? Is the only one way reasonable in here would be to patch `spring-kafka` itself to achieve metrics collection for the case of multiple metric registries in use?

Thanks!

**Comment by garyrussell** [comment, marked as answer]
I think we could change it to detect if one of the registries is marked as `@Primary`.

**Reply by garyrussell** [reply]
https://github.com/spring-projects/spring-kafka/issues/2148

**Reply by garyrussell** [reply]
PR: https://github.com/spring-projects/spring-kafka/pull/2151
